File: MPI C & Fortran Tutorial/F77/timing/mpi_timing.f
C--------------------------------------------------------------------->
C MPI Timing Program - FORTRAN Version
C FILE: mpi_timing.f
C DESCRIPTION:  MPI timing example code.  FORTRAN version.
C   In this example code, an MPI communication timing test is
C   The processor with mypid = 0 will send integer messages of
C   "length" elements to the processor with mypid = 1 "REPS" times.
C   Upon receiving the message a message of identical size is sent
C   back.Before and after timings are made for each rep and an 
C   average calculated when completed.  Also the Bandwidth is 
C   calculated.
C LAST REVISED: 3/15/99 Travis Kerr - for JICS Workshop
C--------------------------------------------------------------------->

      program timing
      include 'mpif.h'

c      IMPLICIT NONE

      external timing_fgettod

      integer REPS,MAXLENGTH
      parameter(REPS = 20)
      parameter(MAXLENGTH = 1000000)

      integer i,n,length
      integer inmsg(MAXLENGTH),outmsg(MAXLENGTH)
      integer mypid,mysize,ierr
      integer status(MPI_STATUS_SIZE)
      
      double precision start,finish,time
      double precision bw

C  *------------------->
C *---> Initialize MPI
C  *------------------->

      call MPI_INIT(ierr)

C  *--------------------------------------------------------->
C *---> Get my rank in the MPI_COMM_WORLD communicator group
C  *--------------------------------------------------------->

      call MPI_COMM_RANK(MPI_COMM_WORLD,mypid,ierr )

C  *---------------------------------------------------------->
C *---> Get the size of the MPI_COMM_WORLD communicator group
C  *---------------------------------------------------------->

      call MPI_COMM_SIZE(MPI_COMM_WORLD,mysize,ierr )

      if (mysize.ne.2) then
       print *, 'Error: Check environment variable MP_PROCS'
       go to 999
      endif

      length = 1

C  *------------------------------>
C *---> Synchronize the processes
C  *------------------------------>

      call MPI_BARRIER(MPI_COMM_WORLD,ierr)

C  *---------------------->
C *---> Task 0 processing
C  *---------------------->

      if (mypid.eq.0) then

       do 15 i=1,4

       time = 0.0e0

C  *--------------------------->
C *---> Round-trip timing test
C  *--------------------------->
       print *, ' '
       print *, ' '
       print *, 'Doing round trip test for:'
       print *, 'Message length = ',length,' integer value(s)'
       print *, 'Message size   = ',4*length,' Bytes'
       print *, 'Number of Reps = ',REPS

       start = MPI_WTIME()

       do 10 n=1,REPS

C  *------------------------------>
C *---> Send message to process 1
C  *------------------------------>

        call MPI_SEND(outmsg(1),length,MPI_INTEGER,1,0,MPI_COMM_WORLD,
     +                ierr)

C  *----------------------------------------------------->
C *---> Now wait to receive the ech reply from process 1
C  *----------------------------------------------------->

        call MPI_RECV(inmsg(1),length,MPI_INTEGER,1,0,MPI_COMM_WORLD,
     +                status,ierr)

   10  continue
        
       finish = MPI_WTIME()
       
C  *--------------------------------------------------------------->
C *---> Calculate round trip time average and bandwidth, and print
C  *--------------------------------------------------------------->

       time = finish - start

       write (*,100) time/REPS
  100  format ('*** Round Trip Avg = ',f8.6,'Sec')

        bw = 2.0*REPS*4.0*length/time

       write (*,110) bw
  110  format ('*** Approximate Bandwidth = ',f18.6,' Byte/Sec')

       write (*,120) bw*8/1000000
  120  format ('                          = ',f12.6,' Megabit/Sec')

       length = 100*length

   15  continue
      endif

C  *------------------------>
C *---> Task one processing
C  *------------------------>

      if (mypid .eq. 1) then
        do 35 i=1,4
        do 40 n=1, REPS

C  *----------------------------------->
C *---> Receive message from process 0
C  *----------------------------------->

         call MPI_RECV(inmsg(1),length,MPI_INTEGER,0,0,MPI_COMM_WORLD,
     +                 status,ierr)

C  *-------------------------------->
C *---> Return message to process 0
C  *-------------------------------->

         call MPI_SEND(outmsg(1),length,MPI_INTEGER,0,0,MPI_COMM_WORLD,
     +                 ierr )
  40    continue   

        length = 100*length

  35    continue

      endif

C  *----------------->
C *---> Finalize MPI
C  *----------------->

999   call MPI_FINALIZE(ierr)
      end

File: MPI C & Fortran Tutorial/C/timing/mpi_timing.c
/**************************************************************************
* MPI Timing Program - C Version
* FILE: mpi_timing.c
* DESCRIPTION:  MPI timing example code.  C version.
*   In this example code, an MPI communication timing test is performed.
*   The processor with mypid = 0 will send integer messages of
*   "length" elements to the processor with mypid = 1 "REPS" times.
*   Upon receiving the message a message of identical size is sent
*   back.Before and after timings are made for each rep and an 
*   average calculated when completed.  Also the Bandwidth is 
*   calculated.
* LAST REVISED: 3/15/99 Travis Kerr - for JICS Workshop
**************************************************************************/

#include "mpi.h"
#include <stdio.h>
#include <sys/time.h>
#include <time.h>
#include <stdlib.h> 

#define	REPS 1000
#define MAXLENGTH 8000000

int main(argc,argv)
int argc;
char *argv[];

{
 int i,n,length;
 int *inmsg, *outmsg;
 int rtt, artt;
 int mypid,mysize;
 int rc;
 double start,finish,time;
 double bw;
 MPI_Status status;

   /*----------------*/
  /* Initialize MPI */
 /*----------------*/

 rc = MPI_Init(&argc,&argv);

   /*-------------------------------------------------------*/
  /* Get the size of the MPI_COMM_WORLD communicator group */
 /*-------------------------------------------------------*/

 rc|= MPI_Comm_size(MPI_COMM_WORLD,&mysize);

   /*------------------------------------------------------*/
  /* Get my rank in the MPI_COMM_WORLD communicator group */
 /*------------------------------------------------------*/

 rc|= MPI_Comm_rank(MPI_COMM_WORLD,&mypid);

 if (mysize != 2)
 {
  fprintf(stderr, "Error: Set environment variable MP_PROCS to 2\n");
  exit(1);
 }

 length = 1;
 inmsg = (int *) malloc(MAXLENGTH*sizeof(int));
 outmsg = (int *) malloc(MAXLENGTH*sizeof(int));

   /*---------------------------*/
  /* synchronize the processes */
 /*---------------------------*/

 rc = MPI_Barrier(MPI_COMM_WORLD);

   /*-------------------*/
  /* Task 0 processing */
 /*-------------------*/

 if (mypid == 0)
 {

    for (i=1; i<=4; i++)
    {

    time = 0.0000000000000;

   /*------------------------*/
  /* round-trip timing test */
 /*------------------------*/

     printf("\n\nDoing round trip test for:\n");
     printf("Message length = %d integer value(s)\n",length);
     printf("Message size   = %d Bytes\n",4*length);
     printf("Number of Reps = %d\n",REPS);


     start = MPI_Wtime();

     for (n=1; n<=REPS; n++)
     {

   /*---------------------------*/
  /* send message to process 1 */
 /*---------------------------*/

      rc = MPI_Send(&outmsg[0],length,MPI_INT,1,0,MPI_COMM_WORLD);

   /*---------------------------------------------------*/
  /* Now wait to receive the echo reply from process 1 */
 /*---------------------------------------------------*/

      rc = MPI_Recv(&inmsg[0],length,MPI_INT,1,0,MPI_COMM_WORLD,&status);

     }

     finish = MPI_Wtime();

   /*------------------------------------------------------------*/
  /* Calculate round trip time average and bandwidth, and print */
 /*------------------------------------------------------------*/

      time = finish - start;

      printf("*** Round Trip Avg = %f uSec\n", time/REPS);

      bw = 2.0*REPS*4.0*length/time;

      printf("*** Bandwidth = %f Byte/Sec\n",bw);
      printf("              = %f Megabit/Sec\n",bw*8.0/1000000.0);

      length = 100*length;

    } 
 }

   /*-------------------*/
  /* Task 1 processing */
 /*-------------------*/

 if (mypid == 1)
 {
    for (i=1; i<=4; i++)
    {

     for (n=1; n<=REPS; n++)
     {

   /*--------------------------------*/
  /* receive message from process 0 */
 /*--------------------------------*/

      rc = MPI_Recv(&inmsg[0],length,MPI_INT,0,0,MPI_COMM_WORLD,&status);

   /*-----------------------------*/
  /* return message to process 0 */
 /*-----------------------------*/

      rc = MPI_Send(&outmsg[0],length,MPI_INT,0,0,MPI_COMM_WORLD);
     }

     length = 100*length;

    }

 }

   /*--------------*/
  /* Finalize MPI */
 /*--------------*/

   MPI_Finalize();
   exit(0);
}
